Update the list of valid ssh keys.

omBratteng opened the pull request:
Update ssh_util.py with latest list of keys (from openssh-8.3p1/sshkey.c),

Added keys:
  rsa-sha2-256-cert-v01@openssh.com
  rsa-sha2-512-cert-v01@openssh.com
  sk-ecdsa-sha2-nistp256-cert-v01@openssh.com
  sk-ecdsa-sha2-nistp256@openssh.com
  sk-ssh-ed25519-cert-v01@openssh.com
  sk-ssh-ed25519@openssh.com
  ssh-xmss-cert-v01@openssh.com
  ssh-xmss@openssh.com
  webauthn-sk-ecdsa-sha2-nistp256@openssh.com

LP: #1877869

omBratteng commented:
I tried testing locally with tox, worked fine. But not sure what the best way to test the newly added keys are.

omBratteng commented:
Thanks @OddBloke, I'll try to get some tests. I'm a bit unfamiliar with tox, is there a way I can run the specific test https://github.com/canonical/cloud-init/blob/master/tests/unittests/test_sshutil.py

OddBloke commented:
Thanks for the update @omBratteng! You've presumably figured this out, but `tox -e py3 -- path/to/test/file.py` will run all the tests in a file, or `tox -e py3 -- path/to/test/file.py::SpecificTestClass` will run a single class in a file. (More generally, anything after the `--` is passed to pytest.)

I notice that the tests only include new cases for a subset of the key types you're adding; is it feasible to expand that list out further, to cover all the newly-added types?

OddBloke commented:
(Oh and as a note to my future self: I have confirmed you've signed the CLA.)

omBratteng commented:
I just spun up a faster VM for running the tests, but I'll try it later.
Yeah, I can try to add tests for the -cert variants too, I just have never used ssh certs. I can also add tests for the rest of the valid SSH keys, as I see it's just a subset of all the valid ones.

OddBloke commented:
On Tue, Jul 14, 2020 at 05:57:01AM -0700, Ole-Martin Bratteng wrote:
> I just spun up a faster VM for running the tests, but I'll try it later.

Fair enough!

> Yeah, I can try to add tests for the -cert variants too, I just have
> never used ssh certs. I can also add tests for the rest of the valid
> SSH keys, as I see it's just a subset of all the valid ones.

Thanks, that would be much appreciated!

omBratteng commented:
@OddBloke added more tests now, and removed some key types which @djmdjm pointed out for me which actually are signatures.
Plus removed `dsa`, `rsa`, `ecdsa` and `ed25519`, as they're not valid public keys, as they're the short names used for human interactions, e.g. "ssh-keygen -t rsa" and they should never appear in public keys

OddBloke commented:
> Plus removed dsa, rsa, ecdsa and ed25519, as they're not valid public keys, as they're the short names used for human interactions, e.g. "ssh-keygen -t rsa" and they should never appear in public keys

Older versions of OpenSSH (I tested with 7.2, the version in Ubuntu 16.04 which we still actively backport cloud-init to) will allow you to SSH in if an `authorized_keys` line starts with "rsa", so removing these is premature.

(I'm running to a meeting, so haven't reviewed anything more than this.)

omBratteng commented:
Gotcha, I can either revert the last three commits, or just f77a723460f73017f65ea8550ffd6010744556b8 which is the one that removes `dsa`, `rsa`, `ecdsa` and `ed25519` from the valid key types, the other two are just testing.

OddBloke commented:
I'd opt for reverting all three commits, if you're happy doing that?

omBratteng commented:
Just pushed a commit that reverts all three

omBratteng commented:
Commits reverted

OddBloke commented on cloudinit/ssh_util.py:
```diff
@@ -17,27 +17,31 @@
 DEF_SSHD_CFG = "/etc/ssh/sshd_config"
 
-# taken from OpenSSH source openssh-7.3p1/sshkey.c:
+# taken from OpenSSH source openssh-8.3p1/sshkey.c:
```
Looking at https://github.com/canonical/cloud-init/pull/487/commits/8b95cc8c1ac75aff91a4d6bb70f8c9527a72e9d9: I'm not an OpenSSH expert, but an examination of the code suggests that this and `rsa-sha2-512` are keys, not certificates:

```c
	{ "rsa-sha2-256", "RSA", NULL, KEY_RSA, 0, 0, 1 },
	{ "rsa-sha2-512", "RSA", NULL, KEY_RSA, 0, 0, 1 },
```

(Compare to:

```c
	{ "ssh-rsa-cert-v01@openssh.com", "RSA-CERT", NULL,
	    KEY_RSA_CERT, 0, 1, 0 },
```
)

However, looking at https://superuser.com/a/1488613 and https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.certkeys suggests that they are not actually key types to be stored on disk, but used exclusively during key exchange.

So: I think excluding these types from the list is correct, but I'd like to see this explanatory comment expanded to include this factor. (Put differently: we are no longer simply "taking" the list from `sshkey.c`, we performing some filtering on it before updating this list, so the comment is inaccurate. I'd like us to capture the filtering for the next person who comes along to update this list.)

OddBloke replied:
(Thinking about it, we should probably also include a pointer to the list in the docs, so the next person also knows to update that list.)

omBratteng replied:
You are correct, they are not stored on the disk. I talked to @djmdjm, whom told me that `webauthn-sk-ecdsa-sha2-nistp256@openssh.com` and the like of `rsa-sha2-256` are not key types, but signature types.

If you take a look at these lines in https://github.com/openssh/openssh-portable/blob/master/sshkey.c#L97-L105
```C
struct keytype {
	const char *name;
	const char *shortname;
	const char *sigalg;
	int type;
	int nid;
	int cert;
	int sigonly;
};
```

the last int signifies signature type, if I'm not 100% wrong. So that's why I removed those in 8b95cc8

I think it's a good idea to leave a comment so that the next person knows to filter out the ones that has the sigonly flag on them

omBratteng commented on tests/unittests/test_sshutil.py:
```diff
@@ -41,24 +41,261 @@
     ),
 }
 
+KEY_TYPES = [
```
That is of possible, I was thinking about doing it. I'll add a commit for that

OddBloke commented:
On Thu, Jul 16, 2020 at 09:31:34AM -0700, Ole-Martin Bratteng wrote:
> > @@ -17,27 +17,31 @@
>  # See: man sshd_config
>  DEF_SSHD_CFG = "/etc/ssh/sshd_config"
>  
> -# taken from OpenSSH source openssh-7.3p1/sshkey.c:
> +# taken from OpenSSH source openssh-8.3p1/sshkey.c:
> 
> You are correct, they are not stored on the disk. I talked to @djmdjm,
> whom told me that `webauthn-sk-ecdsa-sha2-nistp256@openssh.com` and
> the like of `rsa-sha2-256` are not key types, but signature types.
> 
> If you take a look at these lines in
> https://github.com/openssh/openssh-portable/blob/master/sshkey.c#L97-L105
> ```C
> struct keytype {
> 	const char *name;
> 	const char *shortname;
> 	const char *sigalg;
> 	int type;
> 	int nid;
> 	int cert;
> 	int sigonly;
> };
> ```
> 
> the last int signifies signature type, if I'm not 100% wrong. So
> that's why I removed those in 8b95cc8

Aha, yeah, that looks like a much clearer indicator, thanks for pointing
it out!

omBratteng commented:
I've added some comments, both in the source and docs, about where the list comes from. Let me know if you want me to change to wording of it, writing documentation for this sized project isn't something I have done before.

paride commented:
Related PR: https://github.com/canonical/cloud-init/pull/506.

omBratteng commented:
@paride I cherry picked @tsanghan's commit into my branch, updated the `_is_printable_key` list.
And added a pointer in both `cc_ssh_authkey_fingerprints` and `ssh_util`, both pointing to eachother, so the next person knows to update all the lists.

OddBloke commented on cloudinit/config/cc_ssh_authkey_fingerprints.py:
```diff
@@ -57,10 +57,35 @@ def _gen_fingerprint(b64_text, hash_meth='sha256'):
         return '?'
 
 
+# When updating the list, also update the VALID_KEY_TYPES list in
+# cloudinit/ssh_util.py
```
Should we instead import `VALID_KEY_TYPES` from `ssh_util` and use it here?

omBratteng replied:
You thinking of just having the list in `ssh_util`, and then fetching it from there?
That should work, plus, it removes double lists.
I gotta experiment, my python isn't that good

omBratteng replied:
It's now importing `VALID_KEY_TYPES` from `ssh_util`
